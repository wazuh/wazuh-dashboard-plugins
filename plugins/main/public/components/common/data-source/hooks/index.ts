export * from './use-data-source';