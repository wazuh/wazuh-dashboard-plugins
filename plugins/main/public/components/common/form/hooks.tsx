import { useState, useRef } from 'react';
import { isEqual, get, set, cloneDeep } from 'lodash';
import { EpluginSettingType } from '../../../../common/constants';
import {
  CustomSettingType,
  FormConfiguration,
  SettingTypes,
  UseFormReturn,
} from './types';

interface IgetValueFromEventType {
  [key: string]: (event: any) => any;
}

/**
 * Returns the value of the event according to the type of field
 * When the type is not found, it returns the value defined in the default key
 *
 * @param event
 * @param type
 * @returns event value
 */
function getValueFromEvent(
  event: any,
  type: SettingTypes | CustomSettingType,
): any {
  return (getValueFromEventType[type] || getValueFromEventType.default)(event);
}

const getValueFromEventType: IgetValueFromEventType = {
  [EpluginSettingType.switch]: (event: any) => event.target.checked,
  [EpluginSettingType.editor]: (value: any) => value,
  [EpluginSettingType.filepicker]: (value: any) => value,
  [EpluginSettingType.select]: (event: any) => event.target.value,
  [EpluginSettingType.text]: (event: any) => event.target.value,
  [EpluginSettingType.textarea]: (event: any) => event.target.value,
  [EpluginSettingType.number]: (event: any) => event.target.value,
  [EpluginSettingType.password]: (event: any) => event.target.value,
  custom: (event: any) => event.target.value,
  default: (event: any) => event.target.value,
};

export function getFormFields(fields) {
  return Object.entries(fields).reduce(
    (accum, [fieldKey, fieldConfiguration]) => {
      return {
        ...accum,
        [fieldKey]: {
          ...(fieldConfiguration.type === 'arrayOf'
            ? {
                fields: fieldConfiguration.initialValue.map(item =>
                  getFormFields(
                    Object.entries(fieldConfiguration.fields).reduce(
                      (accum, [key]) => ({
                        ...accum,
                        [key]: {
                          initialValue: item[key],
                          currentValue: item[key],
                          defaultValue: fieldConfiguration?.defaultValue,
                        },
                      }),
                      {},
                    ),
                  ),
                ),
              }
            : {
                currentValue: fieldConfiguration.initialValue,
                initialValue: fieldConfiguration.initialValue,
                defaultValue: fieldConfiguration?.defaultValue,
              }),
        },
      };
    },
    {},
  );
}

export function enhanceFormFields(
  formFields,
  {
    fields,
    references,
    setState,
    pathFieldParent = [],
    pathFormFieldParent = [],
  },
) {
  return Object.entries(formFields).reduce(
    (accum, [fieldKey, { currentValue: value, ...restFieldState }]) => {
      // Define the path to fields object
      const pathField = [...pathFieldParent, fieldKey];
      // Define the path to the form fields object
      const pathFormField = [...pathFormFieldParent, fieldKey];
      // Get the field form the fields
      const field = get(fields, pathField);

      return {
        ...accum,
        [fieldKey]: {
          ...(field.type === 'arrayOf'
            ? {
                type: field.type,
                fields: (() => {
                  return restFieldState.fields.map((fieldState, index) =>
                    enhanceFormFields(fieldState, {
                      fields,
                      references,
                      setState,
                      pathFieldParent: [...pathField, 'fields'],
                      pathFormFieldParent: [...pathFormField, 'fields', index],
                    }),
                  );
                })(),
                addNewItem: () => {
                  setState(state => {
                    const _state = get(state, [...pathField, 'fields']);
                    const newstate = set(
                      state,
                      [...pathField, 'fields', _state.length],
                      Object.entries(field.fields).reduce(
                        (accum, [key, { defaultValue }]) => ({
                          ...accum,
                          [key]: {
                            currentValue: cloneDeep(defaultValue),
                            initialValue: cloneDeep(defaultValue),
                            defaultValue: cloneDeep(defaultValue),
                          },
                        }),
                        {},
                      ),
                    );
                    return cloneDeep(newstate);
                  });
                },
              }
            : {
                ...field,
                ...restFieldState,
                type: field.type,
                value,
                changed: !isEqual(restFieldState.initialValue, value),
                error: field?.validate?.(value),
                setInputRef: (reference: any) => {
                  set(references, pathFormField, reference);
                },
                inputRef: get(references, pathFormField),
                onChange: (event: any) => {
                  const inputValue = getValueFromEvent(event, field.type);
                  const currentValue =
                    field?.transformChangedInputValue?.(inputValue) ??
                    inputValue;
                  setState(state => {
                    const newState = set(
                      cloneDeep(state),
                      [...pathFormField, 'currentValue'],
                      currentValue,
                    );
                    return newState;
                  });
                },
              }),
        },
      };
    },
    {},
  );
}

export function mapFormFields(
  {
    formDefinition,
    formState,
    pathFieldFormDefinition = [],
    pathFormState = [],
  },
  callbackFormField,
) {
  return Object.entries(formState).reduce((accum, [key, value]) => {
    const pathField = [...pathFieldFormDefinition, key];
    const fieldDefinition = get(formDefinition, pathField);
    return {
      ...accum,
      [key]:
        fieldDefinition.type === 'arrayOf'
          ? {
              fields: value.fields.map((valueField, index) =>
                mapFormFields(
                  {
                    formDefinition,
                    formState: valueField,
                    pathFieldFormDefinition: [...pathField, 'fields'],
                    pathFormState: [
                      ...[...pathFormState, key],
                      'fields',
                      index,
                    ],
                  },
                  callbackFormField,
                ),
              ),
            }
          : callbackFormField?.(value, key, {
              formDefinition,
              formState,
              pathFieldFormDefinition,
              pathFormState: [...pathFormState, key],
              fieldDefinition,
            }),
    };
  }, {});
}

export const useForm = (fields: FormConfiguration): UseFormReturn => {
  const [formFields, setFormFields] = useState<{
    [key: string]: { currentValue: any; initialValue: any };
  }>(getFormFields(fields));

  const fieldRefs = useRef<{ [key: string]: any }>({});

  const enhanceFields = enhanceFormFields(formFields, {
    fields,
    references: fieldRefs.current,
    setState: setFormFields,
    pathFieldParent: [],
    pathFormFieldParent: [],
  });

  const { changed, errors } = (() => {
    const result = {
      changed: {},
      errors: {},
    };
    mapFormFields(
      {
        formDefinition: fields,
        formState: enhanceFields,
        pathFieldFormDefinition: [],
        pathFormState: [],
      },
      ({ changed, error, value }, _, { pathFormState, fieldDefinition }) => {
        changed &&
          (result.changed[pathFormState] =
            fieldDefinition?.transformChangedOutputValue?.(value) ?? value);
        error && (result.errors[pathFormState] = error);
      },
    );
    return result;
  })();

  function undoChanges() {
    setFormFields(state =>
      mapFormFields(
        {
          formDefinition: fields,
          formState: state,
          pathFieldFormDefinition: [],
          pathFormState: [],
        },
        state => ({ ...state, currentValue: state.initialValue }),
      ),
    );
  }

  function doChanges() {
    setFormFields(state =>
      mapFormFields(
        {
          formDefinition: fields,
          formState: state,
          pathFieldFormDefinition: [],
          pathFormState: [],
        },
        state => ({ ...state, initialValue: state.currentValue }),
      ),
    );
  }

  function forEach(callback) {
    return mapFormFields(
      {
        formDefinition: fields,
        formState: enhanceFields,
        pathFieldFormDefinition: [],
        pathFormState: [],
      },
      callback,
    );
  }

  return {
    fields: enhanceFields,
    changed,
    errors,
    undoChanges,
    doChanges,
    forEach,
  };
};
