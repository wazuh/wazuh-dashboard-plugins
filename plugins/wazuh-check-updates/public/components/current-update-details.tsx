import React from 'react';
import {
  EuiAccordion,
  EuiBadge,
  EuiCallOut,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHeaderLink,
  EuiText,
  EuiSpacer,
} from '@elastic/eui';
import { Update } from '../../common/types';
import { FormattedMessage, I18nProvider } from '@osd/i18n/react';

export interface CurrentUpdateDetailsProps {
  currentUpdate?: Update;
}

export const CurrentUpdateDetails = ({ currentUpdate }: CurrentUpdateDetailsProps) => {
  if (!currentUpdate) {
    return null;
  }

  const currentRelease = `${currentUpdate?.semver.mayor}.${currentUpdate?.semver.minor}`;
  const releaseNotesUrl = `https://documentation.wazuh.com/${currentRelease}/release-notes/release-${currentUpdate.semver.mayor}-${currentUpdate.semver.minor}-${currentUpdate.semver.patch}.html`;
  const upgradeGuideUrl = `https://documentation.wazuh.com/${currentRelease}/upgrade-guide/index.html`;

  return (
    <I18nProvider>
      <EuiCallOut
        title={
          <EuiFlexGroup gutterSize="m">
            <EuiFlexItem grow={false} style={{ maxWidth: 'max-content' }}>
              <FormattedMessage
                id="wazuhCheckUpdates.currentUpdateDetails.title"
                defaultMessage="Wazuh new release is available now!"
              />
            </EuiFlexItem>
            <EuiFlexItem grow={false} style={{ maxWidth: 'max-content' }}>
              <EuiBadge>{currentUpdate.tag}</EuiBadge>
            </EuiFlexItem>
          </EuiFlexGroup>
        }
        color="warning"
        iconType="bell"
      >
        <EuiFlexGroup gutterSize="m">
          <EuiFlexItem grow={false} style={{ maxWidth: 'max-content' }}>
            <EuiHeaderLink
              href={releaseNotesUrl}
              isActive
              target="_blank"
              iconType="popout"
              iconSide="right"
            >
              <FormattedMessage
                id="wazuhCheckUpdates.currentUpdateDetails.releaseNotesLink"
                defaultMessage="Release notes"
              />
            </EuiHeaderLink>
          </EuiFlexItem>
          <EuiFlexItem grow={false} style={{ maxWidth: 'max-content' }}>
            <EuiHeaderLink
              href={upgradeGuideUrl}
              isActive
              target="_blank"
              iconType="popout"
              iconSide="right"
            >
              <FormattedMessage
                id="wazuhCheckUpdates.currentUpdateDetails.upgradeGuideLink"
                defaultMessage="Upgrade guide"
              />
            </EuiHeaderLink>
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer size="m" />
        <EuiAccordion
          id="accordionUpdateDetails"
          buttonContent={
            <FormattedMessage
              id="wazuhCheckUpdates.currentUpdateDetails.showDetails"
              defaultMessage="Show details"
            />
          }
          paddingSize="m"
        >
          <EuiText>
            {currentUpdate.description.split('\r\n').map((line, index) => (
              <p key={index}>{line}</p>
            ))}
          </EuiText>
        </EuiAccordion>
      </EuiCallOut>
    </I18nProvider>
  );
};
