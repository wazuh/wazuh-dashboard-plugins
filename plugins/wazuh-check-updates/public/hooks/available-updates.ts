import { useState, useEffect } from 'react';
import { ApiAvailableUpdates, AvailableUpdates } from '../../common/types';
import { routes } from '../../common/constants';
import { getCore } from '../plugin-services';

export const useAvailableUpdates = () => {
  const [apisAvailableUpdates, setApisAvailableUpdates] = useState<ApiAvailableUpdates[]>([]);
  const [lastCheck, setLastCheck] = useState<string>();

  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const refreshAvailableUpdates = async (forceUpdate = false, returnError = false) => {
    try {
      setIsLoading(true);

      const checkUpdates = sessionStorage.getItem('checkUpdates');
      const alreadyCheckUpdates = checkUpdates === 'executed';

      const response = await getAvailableUpdates(forceUpdate || !alreadyCheckUpdates);

      setApisAvailableUpdates(response?.apis_available_updates || []);
      setLastCheck(response?.last_check_date);
      setError(undefined);

      if (!alreadyCheckUpdates) {
        sessionStorage.setItem('checkUpdates', 'executed');
      }
    } catch (error: any) {
      if (returnError) {
        return error instanceof Error
          ? error
          : typeof error === 'string'
          ? new Error(error)
          : new Error('Error trying to get available updates');
      } else {
        setError(error);
      }
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    refreshAvailableUpdates();
  }, []);

  return { isLoading, apisAvailableUpdates, refreshAvailableUpdates, error, lastCheck };
};

export const getAvailableUpdates = async (forceUpdate = false): Promise<AvailableUpdates> => {
  return await getCore().http.get(routes.checkUpdates, {
    query: {
      checkAvailableUpdates: forceUpdate,
    },
  });
};
