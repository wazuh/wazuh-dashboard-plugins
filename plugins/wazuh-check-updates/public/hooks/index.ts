export { useUserPreferences } from './user-preferences';
