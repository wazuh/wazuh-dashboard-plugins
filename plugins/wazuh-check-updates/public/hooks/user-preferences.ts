import { useState, useEffect } from 'react';
import { UserPreferences } from '../../common/types';
import { getCore } from '../plugin-services';
import { routes } from '../../common/constants';

export const useUserPreferences = () => {
  const [userPreferences, setUserPreferences] = useState<UserPreferences>({});
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const refreshUserPreferences = () => {
    (async () => {
      try {
        setIsLoading(true);
        const response = await getCore().http.get(routes.userPreferences);
        setUserPreferences(response);
        setError(undefined);
      } catch (error: any) {
        setError(error);
      } finally {
        setIsLoading(false);
      }
    })();
  };

  useEffect(() => {
    refreshUserPreferences();
  }, []);

  const updateUserPreferences = async (userPreferences: UserPreferences) => {
    try {
      setIsLoading(true);
      await getCore().http.patch(routes.userPreferences, {
        body: JSON.stringify(userPreferences),
      });
      setUserPreferences(userPreferences);
    } catch (error: any) {
      setError(error);
    } finally {
      setIsLoading(false);
    }
  };

  return {
    isLoading,
    userPreferences,
    updateUserPreferences,
    error,
  };
};
