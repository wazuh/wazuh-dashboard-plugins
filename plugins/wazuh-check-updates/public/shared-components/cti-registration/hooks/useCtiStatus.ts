import { useState, useEffect, useRef, useCallback } from 'react';
import { CtiStatus } from '../types';
import { getApiInfo } from '../../../services/get-api-info';

export const useCtiStatus = () => {
  const [isActive, setIsActive] = useState<CtiStatus>(CtiStatus.INACTIVE);
  const pollingRef = useRef<NodeJS.Timeout | null>(null);
  const previousStatusRef = useRef<CtiStatus | null>(null);

  const stopPolling = useCallback(() => {
    if (pollingRef.current) {
      clearInterval(pollingRef.current);
      pollingRef.current = null;
    }
  }, []);

  const startPolling = useCallback(() => {
    // Avoid multiple intervals
    if (pollingRef.current) {
      clearInterval(pollingRef.current);
    }

    pollingRef.current = setInterval(async () => {
      try {
        const response = await getApiInfo();
        const statusRegistration =
          response.affected_items[0]?.registration?.status ||
          CtiStatus.INACTIVE;

        // Only update isActive if the status changed
        if (previousStatusRef.current !== statusRegistration) {
          setIsActive(statusRegistration);
          previousStatusRef.current = statusRegistration;
        }

        // If the status is no longer PENDING, stop polling
        if (statusRegistration !== CtiStatus.PENDING) {
          stopPolling();
        }
      } catch (error) {
        console.error('Error during polling:', error);
        setIsActive(CtiStatus.ERROR);
        stopPolling();
      }
    }, 5000);
  }, [stopPolling]);

  const checkCtiStatus = useCallback(async () => {
    try {
      const response = await getApiInfo();
      const statusRegistration =
        response.affected_items[0]?.registration?.status || CtiStatus.INACTIVE;

      // Only update isActive if the status changed
      if (previousStatusRef.current !== statusRegistration) {
        setIsActive(statusRegistration);
        previousStatusRef.current = statusRegistration;
      }

      // If PENDING, start polling
      if (statusRegistration === CtiStatus.PENDING) {
        startPolling();
      } else {
        // If not PENDING, stop polling
        stopPolling();
      }
    } catch (error) {
      console.error('Error checking CTI status:', error);
      setIsActive(CtiStatus.ERROR);
      stopPolling();
    }
  }, [startPolling, stopPolling]);

  useEffect(() => {
    checkCtiStatus();
    return () => {
      // Remove polling when component unmounts
      stopPolling();
    };
  }, [checkCtiStatus, stopPolling]);

  return {
    isActive,
    checkCtiStatus,
    startPolling,
    stopPolling,
  };
};
