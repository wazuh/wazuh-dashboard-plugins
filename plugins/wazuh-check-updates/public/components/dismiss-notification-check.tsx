import { EuiCheckbox, EuiIcon, EuiToolTip } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { FormattedMessage, I18nProvider } from '@osd/i18n/react';
import { useUserPreferences } from '../hooks';

export const DismissNotificationCheck = () => {
  const [dismissFutureUpdates, setDismissFutureUpdates] = useState<boolean>();

  const { userPreferences, error, isLoading, updateUserPreferences } = useUserPreferences();

  useEffect(() => {
    if (isLoading) {
      return;
    }
    setDismissFutureUpdates(userPreferences?.hide_update_notifications);
  }, [userPreferences, isLoading]);

  if (error) {
    return null;
  }

  const handleOnChange = (checked: boolean) => {
    updateUserPreferences({ hide_update_notifications: checked });
  };

  return (
    <I18nProvider>
      <EuiCheckbox
        id="check-dismiss"
        label={
          <span>
            <FormattedMessage
              id="wazuhCheckUpdates.dismissNotificationCheck.checkText"
              defaultMessage="Disable updates notifications"
            />{' '}
            <EuiToolTip
              content={
                <FormattedMessage
                  id="wazuhCheckUpdates.dismissNotificationCheck.checkHelp"
                  defaultMessage="This setting determines if a notification will appear every time an update is released"
                />
              }
            >
              <EuiIcon type="questionInCircle" color="primary" />
            </EuiToolTip>
          </span>
        }
        checked={dismissFutureUpdates}
        onChange={(e) => handleOnChange(e.target.checked)}
        disabled={isLoading}
      />
    </I18nProvider>
  );
};
