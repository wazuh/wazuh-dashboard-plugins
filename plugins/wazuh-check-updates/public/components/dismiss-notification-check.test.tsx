import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { CurrentUpdateDetails } from './current-update-details';
import { DismissNotificationCheck } from './dismiss-notification-check';
import { useUserPreferences } from '../hooks';

jest.mock(
  '../../../../node_modules/@elastic/eui/lib/services/accessibility/html_id_generator',
  () => ({
    htmlIdGenerator: () => () => 'htmlId',
  })
);

const mockedUseUserPreferences = useUserPreferences as jest.Mock;
jest.mock('../hooks/user-preferences');

describe('DismissNotificationCheck component', () => {
  test('should render the check', () => {
    mockedUseUserPreferences.mockImplementation(() => ({
      isLoading: false,
      userPreferences: { hide_update_notifications: false, last_dismissed_update: 'v4.2.1' },
    }));

    const { container, getByText } = render(<DismissNotificationCheck />);

    expect(container).toMatchSnapshot();

    const elementWithTag = getByText('Disable updates notifications');
    expect(elementWithTag).toBeInTheDocument();
  });

  test('should return null when there is an error', () => {
    mockedUseUserPreferences.mockImplementation(() => ({
      error: 'Error',
    }));

    const { container } = render(<CurrentUpdateDetails />);

    expect(container).toMatchSnapshot();

    const firstChild = container.firstChild;
    expect(firstChild).toBeNull();
  });
});
