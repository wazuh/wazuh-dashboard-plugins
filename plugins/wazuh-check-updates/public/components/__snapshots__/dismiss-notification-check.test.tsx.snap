// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DismissNotificationCheck component should render the check 1`] = `
<div>
  <div
    class="euiCheckbox"
  >
    <input
      class="euiCheckbox__input"
      id="check-dismiss"
      type="checkbox"
    />
    <div
      class="euiCheckbox__square"
    />
    <label
      class="euiCheckbox__label"
      for="check-dismiss"
    >
      <span>
        Disable updates notifications
         
        <span
          class="euiToolTipAnchor"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--primary euiIcon-isLoading"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
        </span>
      </span>
    </label>
  </div>
</div>
`;

exports[`DismissNotificationCheck component should return null when there is an error 1`] = `<div />`;
